### PR TITLE
Exclude definitions that are not referenced

### PR DESCRIPTION
**Pull request by 0candy**
Connect to strongloop/loopback-component-explorer#143

### CODE REVIEW

File: lib/specgen/type-registry.js [hunk 1]
```diff
@@ -27,11 +28,16 @@ TypeRegistry.prototype.reference = function(typeName) {
   return '#/definitions/' + typeName;
 };
 
+TypeRegistry.prototype.lazyReference = function(name) {
+  this._lazyReference = name;
```
Comment (bajtos) [inline lib/specgen/type-registry.js]: This doesn't look right, should it be `this._lazyReference[name] = true`?

Comment (raymondfeng) [inline lib/specgen/type-registry.js]: I am out of the office until 01/04/2016.

Happy holidays!


Note: This is an automated response to your message  "Re:
[loopback-swagger] WIP - exclude definition that are not referenced (#22)"
sent on 12/21/15 2:34:50 PM.

This is the only notification you will receive while this person is away.

**Comment by bajtos** [comment]
@0candy I am reassigning this PR back to you, so that it stays out of my list of PRs waiting for my review. Please assign it back to me once there is new code to review.

**Comment by 0candy** [comment]
@bajtos Please review

File: test/specgen/model-helper.test.js [hunk 2]
```diff
@@ -5,6 +5,7 @@ var TypeRegistry = require('../../lib/specgen/type-registry');
 var _defaults = require('lodash').defaults;
 var loopback = require('loopback');
 var expect = require('chai').expect;
+var createSwaggerObject = require('loopback-swagger').generateSwaggerSpec;
```
Comment (bajtos) [inline test/specgen/model-helper.test.js]: This is failing the build, you need `require('../..').generateSwaggerSpec` instead. However, since your new test is calling `createSwaggerSpec` instead of `model-helper`, it belongs to a different test file - [test/specgen/swagger-spec-generator.test.js](https://github.com/strongloop/loopback-swagger/blob/248290849a98462b3efa4ac1e5f5203541f30d53/test/specgen/swagger-spec-generator.test.js)

File: test/specgen/model-helper.test.js [hunk 3]
```diff
@@ -189,5 +209,6 @@ function getDefinitionsForModel(modelCtor) {
   var registry = new TypeRegistry();
   modelHelper.registerModelDefinition(modelCtor, registry);
   registry.reference(modelCtor.modelName || modelCtor.definition.name);
-  return registry.getDefinitions();
+  var def = registry.getDefinitions();
+  return def;
```
Comment (bajtos) [inline test/specgen/model-helper.test.js]: Unrelated change, please revert

File: lib/specgen/type-registry.js [hunk 4]
```diff
@@ -18,8 +18,12 @@ function TypeRegistry() {
   });
 }
 
-TypeRegistry.prototype.register = function(typeName, definition) {
-  this._definitions[typeName] = definition;
+TypeRegistry.prototype.register = function(typeName, definitionFunction) {
```
Comment (bajtos) [inline lib/specgen/type-registry.js]: Nitpick: perhaps `definitionOrFactoryFn` would be a better argument name?

File: lib/specgen/type-registry.js [hunk 5]
```diff
@@ -29,19 +33,39 @@ TypeRegistry.prototype.reference = function(typeName) {
...
+      if (this._definitions[name] && !defs[name]) {
+        defs[name] = this._definitions[name]();
+      } else {
+        // https://github.com/strongloop/loopback-explorer/issues/71
+        console.warn('Swagger: skipping unknown type %j.', name);
+      }
     }
+    newRefCount = Object.keys(this._referenced).length;
   }
+  while (currentRefCount !== newRefCount);
```
Comment (bajtos) [inline lib/specgen/type-registry.js]: Coding style - put `while` on the same line as the preceding curly brace.

```js
} while (...);
```

File: lib/specgen/type-registry.js [hunk 6]
```diff
@@ -29,19 +33,39 @@ TypeRegistry.prototype.reference = function(typeName) {
...
+    currentDefCount = Object.keys(this._definitions).length;
+    for (var name in this._definitions) {
+      if (!defs[name]) {
+        defs[name] = this._definitions[name]();
+      }
+    }
+    newDefCount = Object.keys(this._definitions).length;
+  }
+  while (currentDefCount !== newDefCount);
+  return defs;
```
Comment (bajtos) [inline lib/specgen/type-registry.js]: This is duplicating the code in `getDefinitions` and it's quite difficult to spot the difference. Please refactor these two methods to share as much code as possible.

Perhaps you can call `getAllDefinitions` from `getDefinitions`? Something like

```js
var allDefs = this.getAllDefinitions();
var defs = Object.create(null);
for (var name in this._referenced) {
  if (allDefs[name]) {
    defs[name] = allDefs[name];
  } else {
    // https://github.com/strongloop/loopback-explorer/issues/71
    console.warn('Swagger: skipping unknown type %j.', name);
  }
}
```

**Comment by bajtos** [comment]
@0candy reviewed. The changes looks mostly good, I left few minor comments to address.

**Comment by 0candy** [comment]
@bajtos Fix code from comments. PTAL. Thanks!

File: lib/specgen/type-registry.js [hunk 7]
```diff
@@ -27,21 +31,37 @@ TypeRegistry.prototype.reference = function(typeName) {
...
   return defs;
+}
+
+TypeRegistry.prototype.getDefinitions = function() {
+  return getDefinition(true, this);
 };
 
 TypeRegistry.prototype.getAllDefinitions = function() {
-  return _.cloneDeep(this._definitions);
+  return getDefinition(false, this);
```
Comment (bajtos) [inline lib/specgen/type-registry.js]: Boolean arguments are difficult to understand - a person new to this code base would have no idea what is the difference between `getDefinition(false, this)` and `getDefinition(true, this)`.

Also the function name feels misleading to me, I would `getDefinition` to return a single definition.

How about something like this?

```js
getAllDefinitions = function() {
  return this._buildDefinitionsFrom(this._definitions);
}

getDefinitions = function() {
  var defs = this._buildDefinitionsFrom(this._references);
  
  for (var ref in this._references) {
    if (ref in defs) continue;
    // https://github.com/strongloop/loopback-explorer/issues/71
    console.warn('Swagger: skipping unknown type %j.', name);
  }

  return defs;
}
```

**Comment by 0candy** [comment]
@bajtos Fixed. Please review.

**Comment by bajtos** [comment]
Landed, thank you!
